src/emulator/functionsEmulatorRuntime.ts

```diff
@@ -1022,88 +1074,82 @@ async function main(): Promise<void> {
       triggerModule = require(frb.cwd);
     } catch (err) {
       await moduleResolutionDetective(frb, err);
-      return;
+      return undefined;
     }
   }
 
   require("../extractTriggers")(triggerModule, triggerDefinitions);
   triggers = await getEmulatedTriggersFromDefinitions(triggerDefinitions, triggerModule);
 
-  const triggerLogData = { triggers, triggerDefinitions };
-  new EmulatorLog("SYSTEM", "triggers-parsed", "", triggerLogData).log();
+  new EmulatorLog("SYSTEM", "triggers-parsed", "", { triggers, triggerDefinitions }).log();
+  return triggers;
+}
 
-  if (!frb.triggerId) {
-    // This is a purely diagnostic call, it's used as a check to make sure developer code compiles and runs as
-    // expected, so we don't have any function to invoke.
-    return;
-  }
+async function flushAndExit(code: number) {
+  await EmulatorLog.waitForFlush();
+  process.exit(code);
+}
 
-  if (!triggers[frb.triggerId]) {
-    new EmulatorLog(
-      "FATAL",
-      "runtime-status",
-      `Could not find trigger "${frb.triggerId}" in your functions directory.`
-    ).log();
-    return;
-  } else {
-    logDebug(`Trigger "${frb.triggerId}" has been found, beginning invocation!`);
-  }
+async function main(): Promise<void> {
+  logDebug("Functions runtime initialized.", {
+    cwd: process.cwd(),
+    node_version: process.versions.node,
+  });
 
-  const trigger = triggers[frb.triggerId];
-  logDebug("", trigger.definition);
-  const mode = trigger.definition.httpsTrigger ? "HTTPS" : "BACKGROUND";
+  process.on("message", async (message: string) => {
```

Comment:
I should point out: `.on` does _not_ take an asynchronous function. What this will do is start this function, but another message could show up and possibly(?) overrun the execution of the first callback. If this is dependent on this being serial, I don't think that would be guaranteed by this. (also, this didn't really dawn on me until recently - there's a lint error that complains about this... I don't know the best way to mix promises and event emitters off the top of my head though - there's probably an observable library that does it... this is just food for thought)

Reply:
If you want this to be serial, just keep a reference to the previous promise and chain it, like `run = run.then(() => { ... })`. (And with proper error handling so that if previous run failed we can still move on.) If you don't actually want it to be serial and you just want the nice async await goodies, then just make another function (e.g. `handleTrigger()`) and explicitly call it in the event handler with a nice comment:

```
process.on("message", () => {
  // Execute the function async. All error handling is done within so we can safely ignore the returned promise here.
  handleTrigger();
});
```

I used to mess with those things on the Node.js development and these patterns had saved us a lot. Either way, putting async on event listeners is confusing and making a separate function makes the intent clear here.

Reply:
Thanks for pointing this out!  I moved to a sync handling model like Yuchen described, PTAL.